### PR TITLE
Add Narayana JMS connection and session pooling

### PR DESCRIPTION
**Pull request by nielsm5**
Spring boot seems to not use the ConnectionFactoryProxy, see https://github.com/snowdrop/narayana-spring-boot/commit/dbe74dc13ed4ef936db839bc7851ef82fcdeedfe
https://github.com/snowdrop/narayana-spring-boot

### CODE REVIEW

File: core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaConnectionFactoryFactory.java [hunk 1]
```diff
@@ -20,25 +20,60 @@
...
+
 	@Override
 	protected ConnectionFactory augment(ConnectionFactory connectionFactory, String connectionFactoryName) {
 		if (connectionFactory instanceof XAConnectionFactory) {
 			XAResourceRecoveryHelper recoveryHelper = new JmsXAResourceRecoveryHelper((XAConnectionFactory) connectionFactory);
 			this.transactionManager.registerXAResourceRecoveryHelper(recoveryHelper);
-			return new ConnectionFactoryProxy((XAConnectionFactory) connectionFactory, transactionManager.getTransactionHelper());
+
+			if(maxConnectionPoolSize > 1) {
+				return createConnectionFactoryPool(connectionFactory);
```
Comment (tnleeuw) [inline core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaConnectionFactoryFactory.java:49]: Is the custom`NarayanaTransactionHelper` not needed for proper release of sessions / connections in this case?

File: core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaConnectionFactoryFactory.java [hunk 2]
```diff
@@ -20,25 +20,60 @@
...
+		JmsPoolXAConnectionFactory pooledConnectionFactory = new JmsPoolXAConnectionFactory();
+		pooledConnectionFactory.setTransactionManager(this.transactionManager.getTransactionManager());
+		pooledConnectionFactory.setConnectionFactory(xaConnectionFactory);
+
+		pooledConnectionFactory.setMaxConnections(maxConnectionPoolSize);
+		pooledConnectionFactory.setConnectionIdleTimeout(getMaxIdleTime() * 1000);
+		pooledConnectionFactory.setConnectionCheckInterval(connectionCheckInterval);
+		pooledConnectionFactory.setUseProviderJMSContext(false); // indicates whether the pool should include JMSContext in the pooling, when set to true, it disables connection pooling
+
+		pooledConnectionFactory.setMaxSessionsPerConnection(maxSessionPoolSize); // defaults to 500
```
Comment (tnleeuw) [inline core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaConnectionFactoryFactory.java:70]: 500 is not the default we set, I find your comment confusing in this way.

File: core/src/main/resources/AppConstants.properties [hunk 3]
```diff
@@ -106,15 +106,44 @@ transactionmanager.narayana.stuckTimerTime=30000
...
+# Maximum number of physical connections that you can create in this pool.
+transactionmanager.narayana.jdbc.connection.maxPoolSize=20
+
+## JMS Connection Pool properties for Narayana
+# Maximum number of physical connections that you can create in this pool.
+transactionmanager.narayana.jms.connection.maxPoolSize=20
+# Amount of time a connection can be unused or idle until it can be discarded.
+transactionmanager.narayana.jms.connection.maxIdleTime=60
+# Maximum number of jms sessions per connection that can be created in the connection pool.
+transactionmanager.narayana.jms.session.maxPoolSize=20
```
Comment (tnleeuw) [inline core/src/main/resources/AppConstants.properties:144]: "max nr of sessions per connection" ≠ "session pool size"; there is no pool of sessions. I think it is better to change this property name, for instance `transactionmanager.narayana.jms.connection.maxSessions=500`

**Review by tnleeuw — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaConnectionFactoryFactory.java (hunk 1 above)
Comment (nielsm5) [reply core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaConnectionFactoryFactory.java:49]: I mentioned this in the description of this PR, but simply put, no. In Spring Boot it used to be like that, until they removed it for some unknown reason.

**Review by tnleeuw — APPROVED** [review]

**Review by gvanbrakel — APPROVED** [review]

**Pull request merged**